### PR TITLE
[AD-196] Add handling for nulls in SUM

### PR DESCRIPTION
**Pull request by andiemontoyeah**
### Summary
- SUM is returning 0 when all values are NULL but this case should return NULL 
- Added additional logic to determine if there are any non-null values before returning sum. Fix works for SUM and SUM(DISTINCT). 

### Description 
- Added logic to push values to sum in an array and then determine if there are any non-null values before returning sum. 
- Added similar fix to COUNT(DISTINCT).
- Tweaked COUNT to use greater than ($gt) NULL rather than $cond and $ifNull since output was incorrect for boolean columns.
- Updated unit tests and added 1 statement test 

### Related Issue
https://bitquill.atlassian.net/browse/AD-196 

### Additional Reviewers
@affonsoBQ
@andiem-bq
@alexey-temnikov
@birschick-bq

### CODE REVIEW

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java [hunk 1]
```diff
@@ -166,7 +163,9 @@ public DocumentDbAggregate(final RelOptCluster cluster, final RelTraitSet traitS
                 "{$group: " + Util.toString(list, "{", ", ", "}") + "}");
         final List<String> fixups = getFixups(aggCalls, groupSet, inNames, outNames);
 
-        if (!groupSet.isEmpty() || aggCalls.stream().anyMatch(AggregateCall::isDistinct)) {
+        if (!groupSet.isEmpty()
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:166]: Will need a project stage after the group stage  if we have a non-empty group set (has GROUP BY clause), are using a distinct aggregation call (DISTINCT sum/count/etc) or are using SUM

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java [hunk 2]
```diff
@@ -261,24 +276,13 @@ private static String setToAggregate(final SqlAggFunction aggFunction, final Str
             final ImmutableBitSet groupSet,
             final List<String> inNames,
             final List<String> outNames) {
-        final List<String> fixups;
+        // DocumentDB: modified - start
+        final List<String> fixups = new ArrayList<>();
+        int i = 0;
         if (groupSet.cardinality() == 1) {
-            fixups = new AbstractList<String>() {
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:266]: This was a bug from when we added support for distinct aggregation calls. If there were any distinct aggregate calls combined with a group set of 1, the output would have been wrong. :)

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java [hunk 3]
```diff
@@ -287,20 +291,24 @@ private static String setToAggregate(final SqlAggFunction aggFunction, final Str
                                 + maybeQuote("$_id." + acceptedMongoFieldName(inNames.get(group))));
                 ++i;
             }
-            for (AggregateCall aggCall : aggCalls) {
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:290]: Moved loop out of else block. This was a bug from when we added support for distinct aggregation calls. If there were any distinct aggregate calls combined with a group set of 1 (the if block), the output would have been wrong. :)

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java [hunk 4]
```diff
@@ -201,19 +200,20 @@ private static String toMongo(final SqlAggFunction aggregation, final List<Strin
...
         }
 
         if (aggregation == SqlStdOperatorTable.COUNT) {
             if (args.isEmpty()) {
                 return "{$sum: 1}";
             } else {
                 final String inName = inNames.get(args.get(0));
-                return "{$sum: {$cond: [ {$ifNull: " +
-                        "[" + DocumentDbRules.quote("$" + inName) + ", false]}, 1, 0]}}";
+                return "{$sum: {$cond: [ {$gt: " + "[" + maybeQuote("$" + inName) + ", null]}, 1, 0]}}";
```
Comment (andiemontoyeah) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:210]: the $ifNull operator would set any null values to "false" and then we would count the non-false values. This was giving a wrong output for boolean columns where we want to consider false values too. Changed to use the > null check which we use in other places to check for nulls.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 3 above)
Comment (birschick-bq) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:290]: Did you add a test for this bug?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 1 above)
Comment (alexey-temnikov) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:166]: Do we have any issues in handling nulls in other aggregate functions?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 2 above)
Comment (alexey-temnikov) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:266]: Should we rename `i` to make it easier to read/understand logic? WOuld it be better to use `groupSet.length()` instead?

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 1 above)
Comment (andiemontoyeah) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:166]: No $min/$max/$avg already correctly return null for this case and then we fixed COUNT before.

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 2 above)
Comment (andiemontoyeah) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:266]: renamed i. 👍  but kept cardinality(). groupset is a bit set here. length() will return index of highest set bit (i think?) while cardinality returns # of set bit

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java (hunk 3 above)
Comment (andiemontoyeah) [reply calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java:290]: added 👍

File: calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java [hunk 5]
```diff
@@ -166,7 +163,9 @@ public DocumentDbAggregate(final RelOptCluster cluster, final RelTraitSet traitS
                 "{$group: " + Util.toString(list, "{", ", ", "}") + "}");
         final List<String> fixups = getFixups(aggCalls, groupSet, inNames, outNames);
 
-        if (!groupSet.isEmpty() || aggCalls.stream().anyMatch(AggregateCall::isDistinct)) {
+        if (!groupSet.isEmpty()
+                || aggCalls.stream().anyMatch(AggregateCall::isDistinct)
+                || aggCalls.stream().anyMatch(aggregateCall -> aggregateCall.getAggregation() == SqlStdOperatorTable.SUM)) {
```
Comment (birschick-bq) [inline calcite-adapter/src/main/java/software/amazon/documentdb/jdbc/calcite/adapter/DocumentDbAggregate.java]: Would this be more efficient if you OR'd the two condition inside one `anyMatch()` call?

**Review by birschick-bq — COMMENTED** [review]
Small change requested.

**Review by birschick-bq — APPROVED** [review]

**Review by alexey-temnikov — APPROVED** [review]

**Review by affonsov — APPROVED** [review]

**Pull request merged**